### PR TITLE
Properly identify camel cased acronyms as UpperCamelCase

### LINKED ISSUE #6305: The IncorrectCase diagnostic is overzealous for types that should be Camel Cased
For the following `lib.rs`:
```rust
pub struct AxisAlignedBoundingBox; // A
pub struct AABB; // B
pub struct A_A_B_B; // C
```
`cargo check` will only mark version "C" as having incorrect case:
```sh
$ cargo check
warning: type `A_A_B_B` should have an upper camel case name
 --> src/lib.rs:3:12
  |
3 | pub struct A_A_B_B; // C
  |            ^^^^^^^ help: convert the identifier to upper camel case: `AABB`
```

While rust-analyzer will erroneously mark version "B" as having incorrect case, saying 
> Hint [rust-analyzer] Structure `AABB` should have CamelCase name, e.g. `Aabb`

I agree with cargo here. I'd say `AABB` isn't SCREAMING_SNAKE_CASE since there are no underscores.

If you guys agree that this is a bug, I'd like to open a PR for this.

### PR DESCRIPTION
**Pull request by ArifRoktim**
This closes #6305.

### CODE REVIEW

**Comment by ArifRoktim** [comment]
Realized that the following doesn't work as expected:
```rust
#[allow(non_camel_case_types)]
pub struct FOO_BAR;
```
```
[rust-analyzer] [H] Structure `FOO_BAR` should have CamelCase name, e.g. `FooBar`
```
I don't know how long it'll take to fix that, so feel free to merge away.

**Comment by popzxc** [comment]
Yeah, that was discussed here: https://github.com/rust-analyzer/rust-analyzer/pull/6130#discussion_r500501884

TL;DR rust analyzer doesn't currently collect information about attributes applied to items. This is indeed a complex task due to variety of places where `allow` attribute may appear, so for now it is expected that users who don't want rust analyzer to complain will just disable this diagnostic selectively.

The PR itself LGTM.

**Comment by matklad** [comment]
bors r=popzxc
